libraries/AP_VideoTX/AP_VideoTX.cpp

```diff
@@ -366,96 +503,25 @@ void AP_VideoTX::change_power(int8_t position)
     if (position < 0 || position > 5) {
         return;
     }
-
-    uint16_t power = 0;
-    // 0 or 25
-    if (_max_power_mw < 100) {
-        switch (position) {
-            case 3:
-            case 4:
-            case 5:
-                power = 25;
-                break;
-            default:
-                power = 0;
-                break;
-        }
-    }
-    // 0, 25 or 100
-    else if (_max_power_mw < 200) {
-        switch (position) {
-            case 0:
-                power = 0;
-                break;
-            case 5:
-                power = 100;
-                break;
-            default:
-                power = 25;
-                break;
+    // first find out how many possible levels there are
+    uint8_t num_active_levels = 0;
+    for (uint8_t i = 0; i < VTX_MAX_POWER_LEVELS; i++) {
+        if (_power_levels[i].active != PowerActive::Inactive) {
+            num_active_levels++;
         }
     }
-    // 0, 25, 100 or 200
-    else if (_max_power_mw < 500) {
-        switch (position) {
-            case 1:
-            case 2:
-                power = 25;
-                break;
-            case 3:
-            case 4:
-                power = 100;
-                break;
-            case 5:
-                power = 200;
-                break;
-            default:
-                power = 0;
-                break;
-        }
-    }
-    // 0, 25, 100, 200 or 500
-    else if (_max_power_mw < 800) {
-        switch (position) {
-            case 1:
-            case 2:
-                power = 25;
-                break;
-            case 3:
-                power = 100;
-                break;
-            case 4:
-                power = 200;
-                break;
-            case 5:
-                power = 500;
-                break;
-            default:
-                power = 0;
-                break;
+    // iterate through to find the level
+    uint16_t level = constrain_int16(roundf((num_active_levels * (position + 1)/ 6.0f) - 1), 0, num_active_levels - 1);
+    debug("looking for pos %d power level %d from %d", position, level, num_active_levels);
+    uint16_t power = 0;
+    for (uint8_t i = 0, j = 0; i < num_active_levels; i++, j++) {
+        while (_power_levels[j].active == PowerActive::Inactive && j < VTX_MAX_POWER_LEVELS-1) {
```

Comment:
the j < VTX_MAX_POWER-1 needs to be first in the while test

Reply:
Fixed